soccer/gameplay/skills/goalside_mark.py

```diff
@@ -0,0 +1,194 @@
+                mark_pos -
+                shot_pt).normalized() * self.ratio * mark_line.length()
+
+        #Drawing for simulator
+        main.system_state().draw_circle(mark_pos, constants.Robot.Radius * 1.2,
+                                        (0, 127, 255), "Mark")
+
+        #Move robot into position and face the ball
+        self.robot.move_to(self._target_point)
+        self.robot.face(main.ball().pos)
+
+    #Ratio of distance to mark point v. distance to ball for target point
+    @property
+    def ratio(self):
+        return self._ratio
+
+    @ratio.setter
+    def ratio(self, value):
+        self._ratio = min(max(value, 0.0), 1.0)
+
+    #Max distance for inclusion of ratio in target point
+    #If robot distance exceeds this then it will target the closest point on the mark line
+    @property
+    def mark_line_thresh(self):
+        return self._mark_line_thresh
+
+    @mark_line_thresh.setter
+    def mark_line_thresh(self, value):
+        self._mark_line_thresh = value
+
+    # Overrides mark_robot with a static point
+    @property
+    def mark_point(self) -> robocup.Point:
+        return self._mark_point
+
+    @mark_point.setter
+    def mark_point(self, value: robocup.Point):
+        self._mark_point = value
+
+    #Robot to mark (what it sounds like)
+    @property
+    def mark_robot(self) -> robocup.Robot:
+        return self._mark_robot
+
+    @mark_robot.setter
+    def mark_robot(self, value: robocup.Robot):
+        self._mark_robot = value
+
+    # Choose a robot close to the mark point
+    def role_requirements(self):
+        req = super().role_requirements()
+        if self._target_point is not None:
+            req.destination_shape = self._target_point
+        return req
+
+    def get_mark_line(self):
+        #Finds the line segment between the ball and the highest danger shot point
+        #Cuts off the portion of the line that is inside of the goal box
+        #@return: LineSegment to defend on, shot_point
+
+        #Define the segments where the defender can go closest the goal
+        offset = constants.Robot.Radius
+        left_seg = robocup.Segment(
+            robocup.Point(-constants.Field.PenaltyLongDist / 2 - offset, 0),
+            robocup.Point(-constants.Field.PenaltyLongDist / 2 - offset,
+                          constants.Field.PenaltyShortDist + offset))
+        right_seg = robocup.Segment(
+            robocup.Point(constants.Field.PenaltyLongDist / 2 + offset, 0),
+            robocup.Point(constants.Field.PenaltyLongDist / 2 + offset,
+                          constants.Field.PenaltyShortDist + offset))
+        top_seg = robocup.Segment(
+            robocup.Point(-constants.Field.PenaltyLongDist / 2,
+                          constants.Field.PenaltyShortDist + offset),
+            robocup.Point(constants.Field.PenaltyLongDist / 2,
+                          constants.Field.PenaltyShortDist + offset))
```

Comment:
I see why this is here, but I wish there was a better way to do this (Intersection of a segment/line with the goal box with some offset). If you are comfortable making some other changes to the C++ interface to allow for this that would be awesome. If not, I can add it sometime.

Basically, the [field](https://github.com/RoboJackets/robocup-software/blob/master/soccer/gameplay/constants.py#L34) in python refers back to it's C++ [counterpart](https://github.com/RoboJackets/robocup-software/blob/f0cf06638ee427563ca7162bbbd65809c30f59f2/common/Field_Dimensions.hpp#L47). Here, there is something called our goal zone shape which is a composite shape. In short a [composite shape](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/CompositeShape.hpp) is a list of shapes which are or'ed together. In old times (like a year or two ago), the defensive area was two arc's with a rectangle in the middle so this made sense. When we redid it for rectangles, we left the composite shape wrapper for some reason. 

Anyways, now, the composite shape just contains a single [rectangle](https://github.com/RoboJackets/robocup-software/blob/f0cf06638ee427563ca7162bbbd65809c30f59f2/common/Field_Dimensions.hpp#L127). The [rectangle](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/Rect.hpp) class doesn't allow for segment intersection, but that's a good thing to add in my mind. If we overload the `bool intersects` function and take in a segment, very close to how it's done in the [segment `intersect` function](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/Segment.hpp#L51), I think it'll work out pretty well. In terms of actual implementation, you should be able to follow the same way they do the [`nearPoint` function](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/Rect.cpp#L105) where they create the 4 segments and just call the segment functions. It's up to you, but I think returning both intersections would be the best bet and let the user figure out which to use (and maybe return the number of intersect points). Finally, to access in the python side, you'll have to add the function call to [this location in robocup-py](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/Rect.cpp#L105), but you'll have to follow the setup in the [segment intersect function](https://github.com/RoboJackets/robocup-software/blob/master/soccer/gameplay/robocup-py.cpp#L683) (and [here](https://github.com/RoboJackets/robocup-software/blob/master/soccer/gameplay/robocup-py.cpp#L242)) to fully deal with the weird python <-> C++ conversion issues. A test or two should be added into [this file](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/RectTest.cpp) also.

In the future, composite shape should probably be removed, and just a rectangle should be used instead, but that may be out of scope for this.

Reply:
Oh, and to add the offsets, you can use the [`expand`](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/Rect.hpp#L67) function of rectangles with the [mixx/maxx/miny/maxy](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/Rect.hpp#L73) to create the X meter offsets outside the given rectangle.

Reply:
I'll give this a go - but expect to do a pretty heavy review on my C++ skills.